src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs

```diff
@@ -0,0 +1,124 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+
+namespace System.Diagnostics
+{
+    public sealed class ActivitySource : IDisposable
+    {
+        private static object s_syncObject = new object();
+        private static List<ActivitySource> s_activeSources = new List<ActivitySource>();
+        private ActivitySource() { throw new InvalidOperationException(); }
+
+        /// <summary>
+        /// Event to subscribe to get the notification when any ActivitySource object get created or released.
+        /// </summary>
+        public static event EventHandler<ActivitySourceEventArgs>? OperationEvent;
+
+        /// <summary>
+        /// Event to subscribe to get the notification when any Activity object get created or disposed.
+        /// </summary>
+        public event EventHandler<ActivitySourceEventArgs>? ActivityEvent;
+
+        /// <summary>
+        /// Construct an ActivitySource object with the input name
+        /// </summary>
+        /// <param name="name">The name of the ActivitySource object</param>
+        public ActivitySource(string name)
+        {
+            if (name == null)
+            {
+                throw new ArgumentNullException(nameof(name));
+            }
+
+            lock (s_syncObject)
```

Comment:
These instances are created rarely enough that this won't cause significant contention?

Reply:
Yes, We are not expecting ActivitySource will be created much.  

---
In reply to: [383014589](https://github.com/dotnet/runtime/pull/32660#discussion_r383014589) [](ancestors = 383014589)